bash-preexec.sh

```diff
@@ -70,15 +71,25 @@ __bp_require_not_readonly() {
   done
 }
 
-# Remove ignorespace and or replace ignoreboth from HISTCONTROL
-# so we can accurately invoke preexec with a command from our
-# history even if it starts with a space.
+# Remove ignorespace and or replace ignoreboth from HISTCONTROL so we can
+# accurately invoke preexec with a command from our history even if it starts
+# with a space.  We then remove commands that start with a space from the
+# history "manually", if either "ignorespace" or "ignoreboth" was part of
+# HISTCONTROL.
 __bp_adjust_histcontrol() {
     local histcontrol
-    histcontrol="${HISTCONTROL//ignorespace}"
+    if [[ "$HISTCONTROL" == *"ignorespace"* || "$HISTCONTROL" == *"ignoreboth"* ]]; then
+      __bp_ignorespace=yes
+    fi
+    histcontrol="${HISTCONTROL//ignorespace:}"
+    histcontrol="${histcontrol//:ignorespace}"
+    histcontrol="${histcontrol//ignorespace}"
```

Comment:
We can perform adjustments of `ignoreboth` and `ignoredups` at once.

```suggestion
    histcontrol=:${histcontrol//:/::}:
    histcontrol=${histcontrol//:ignorespace:}
    histcontrol=${histcontrol//:ignoreboth:/:ignoredups:}
    histcontrol=${histcontrol//::/:}
    histcontrol=${histcontrol#:}
    histcontrol=${histcontrol%:}
```

Reply:
Applied.